packages/workbox-routing/lib/Router.mjs

```diff
@@ -0,0 +1,328 @@
+ *
+ * You can use the `handleRequest()` method to pass a `FetchEvent` through the
+ * router and ultimately get a "routed response" back.
+ * If you'd like this to be handled automatically for you, calling
+ * `addFetchListener()` will cause the `Router` to respond to `fetch` events.
+ *
+ * If a request matches multiple routes, precedence will be given to the
+ * **earliest** registered route.
+ *
+ * @memberof module:workbox-routing
+ */
+class Router {
+  /**
+   * Constructs a new `Router` instance, without any registered routes.
+   */
+  constructor() {
+    // _routes will contain a mapping of HTTP method name ('GET', etc.) to an
+    // array of all the corresponding Route instances that are registered.
+    this._routes = new Map();
+    this._isListenerRegistered = false;
+  }
+
+  /**
+   * This will register a `fetch` event listener on your behalf which will check
+   * the incoming request to see if there's a matching route, and only respond
+   * if there is a match.
+   *
+   * @return {boolean} Returns `true` if this is the first time the method is
+   * called and the listener was registered. Returns `false` if called again,
+   * as the listener will only be registered once.
+   */
+  addFetchListener() {
+    if (!this._isListenerRegistered) {
+      this._isListenerRegistered = true;
+      self.addEventListener('fetch', (event) => {
+        const responsePromise = this.handleRequest(event);
+        if (responsePromise) {
+          event.respondWith(responsePromise);
+        }
+      });
+      return true;
+    } else {
+      _private.logger.warn(
+        `addFetchListener() has already been called for this Router.`);
+      return false;
+    }
+  }
+
+  /**
+   * This can be used to apply the routing rules to generate a response for a
+   * given request inside your own `fetch` event handler.
+   *
+   * @param {FetchEvent} event The event passed in to a `fetch` handler.
+   * @return {Promise<Response>|undefined} Returns a promise for a response,
+   * taking the registered routes into account. If there was no matching route
+   * and there's no `defaultHandler`, then returns undefined.
+   */
+  handleRequest(event) {
+    if (process.env.NODE_ENV !== 'production') {
+      core.assert.isInstance(event, FetchEvent, {
+        moduleName: 'workbox-routing',
+        className: 'Router',
+        funcName: 'handleRequest',
+        paramName: 'event',
+      });
+    }
+
+    const url = new URL(event.request.url);
+    if (!url.protocol.startsWith('http')) {
+      _private.logger.log(`The URL '${url}' does not start with 'http', so ` +
+        `it can't be handled.`);
+      return;
+    }
+
+    let {handler, params} = this._findHandlerAndParams(event, url);
+
+    // If we don't have a handler because there was no matching route, then
+    // fall back to defaultHandler if that's defined.
+    if (!handler && this.defaultHandler) {
+      handler = this.defaultHandler;
+    }
+
+    if (handler) {
```

Comment:
Nit: I'd invert this to a !handler check and return undefined early, then un-nest the following logic.

Reply:
I'm going to keep the logic as-is, as it reflects there being a few possible states:
- `handler` is set by a matching route
- `handler` is set by `defaultHandler`
  - `catchHandler` is set
  - `catchhandler` isn't set
- `handler` isn't set

Reply:
ok. It's a personal preference. I always favor early return than needless nesting.